contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -31,47 +30,135 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
-                strategyStorage.farmId = i;
-                isPoolFound = true;
-                break;
-            }
-        }
+    function reinitialize(
+        ITraderJoeLBPair lbPair,
+        ITraderJoeLBRouter lbRouter,
+        uint256 binStep,
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocationsX,
+        uint256[] calldata binAllocationsY,
+        NameValuePair[] calldata params
+    ) external reinitializer(2) {
+        checkBinIds(binIds);
+        checkBinAllocations(binAllocationsX);
+        checkBinAllocations(binAllocationsY);
 
-        if (!isPoolFound) {
-            revert InvalidTraderJoeLpToken();
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
+            .getStorage();
+
+        // Initialization.
+        strategyStorage.lbPair = lbPair;
+        strategyStorage.lbRouter = lbRouter;
+        strategyStorage.tokenX = IERC20Upgradeable(lbPair.tokenX());
+        strategyStorage.tokenY = IERC20Upgradeable(lbPair.tokenY());
+        strategyStorage.binStep = binStep;
+        strategyStorage.binIds = binIds;
+        strategyStorage.binAllocationsX = binAllocationsX;
+        strategyStorage.binAllocationsY = binAllocationsY;
+
+        if (strategyStorage.tokenX == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenY
+            );
+        } else if (strategyStorage.tokenY == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenX
+            );
+        } else {
+            revert InvalidTraderJoeLBPair();
         }
+
+        // Migration from V1 to V2 consists of three steps.
+        // 0. Withdraw all depositToken and pairDepositToken from V1.
+        // 1. Swap all pairDepositToken to depositToken.
+        // 2. Deposit all depositToken into V2.
+
+        // 0. Withdraw all depositToken and pairDepositToken from V1.
+        uint256 depositTokenBefore = depositToken.balanceOf(address(this));
+
+        uint256 lpBalanceToWithdraw = strategyStorage
+            .masterChef
+            .userInfo(strategyStorage.farmId, address(this))
+            .amount;
+
+        strategyStorage.masterChef.withdraw(
+            strategyStorage.farmId,
+            lpBalanceToWithdraw
+        );
+
+        strategyStorage.lpToken.approve(
+            address(strategyStorage.router),
+            lpBalanceToWithdraw
+        );
+
+        strategyStorage.router.removeLiquidity(
+            address(strategyStorage.pairDepositToken),
+            address(depositToken),
+            lpBalanceToWithdraw,
+            0,
+            0,
+            address(this),
+            // solhint-disable-next-line not-rely-on-time
+            block.timestamp
+        );
+
+        // 1. Swap all pairDepositToken to depositToken.
+        address[] memory path = new address[](2);
```

Comment:
how much does this swap and the follow up swap cost us? Because here we swap pairDepositToken->depositToken, and later in the _deposit() we do the exact same swap, but in reverse order. If it cost too much we might want to pass a parameter through the params variable to _deposit to skip swapping

Reply:
In the worst case, the swap cost is around $10.8.

Currently, Calm portfolio has around $36000 TVL and TraderJoe has 10% of it. Let's assume that we make a swap to the maximum. Note that TraderJoe's swap fee is 0.3%.

i) Swap withdrawn USDC.e to USDC is $5.4
36000 * 0.1 * 0.5 * 0.003 = 5.4 (Currently, V1 pool asset ratio is around 1:1)

ii) Swap it back to USDCe is also $5.4

iii) Price impact by us would be negligible in that our swap amount is 0.6% of the pool TVL.
36,000 / 5,583,136 = 0.00644798908

iv) Currently, V2 pool has APR 1.98% to we make around $0.19 daily. So it will take around 55 days to make up the loss.
36000 * 0.1 * 0.0198 / 365 = 0.19528767123
10.8 / 0.19528767123 = 55.3030303038

But my time cost is much expensive than that, so if you're worry about that then let's just put $10.8 from our treasury.

Reply:
But I'll think about minimizing swap fee during adjusting bins though

Reply:
1. my gut feeing is that 0.6% of the pool is quite substantial when it comes to slippage, why do you think it will cause minimal slippage?
2. 'But I'll think about minimizing swap fee during adjusting bins though' - can you explain what you mean please?